Refactor railsrc file location to be xdg compliant

quintrino opened the pull request:
### Summary

The XDG Base Directory Specification (which is currently used by
FOSS projects such as Git, Tmux, Pry, Rspec) provides a default
location for various file formats, including config/rc files.

This comment refactors app_generator.rb to load railsrc from
XDG_CONFIG_HOME if both XDG_CONFIG_HOME is set and rails/railrc
exists within the XDG_CONFIG_HOME location.

To maintain backwards compatibility it defaults back to ~/.railsrc
if either XDG_CONFIG_HOME is not set or there is no rails/railsrc.

### Other Information

<!-- If there's anything else that's important and relevant to your pull
request, mention that information here. This could include
benchmarks, or other information.

If you are updating any of the CHANGELOG files or are asked to update the
CHANGELOG files by reviewers, please add the CHANGELOG entry at the top of the file.

Finally, if your pull request affects documentation or any non-code
changes, guidelines for those changes are [available
here](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#contributing-to-the-rails-documentation)

Thanks for contributing to Rails! -->

quintrino committed:
Refactor railsrc file location to be xdg compliant

The XDG Base Directory Specification (which is currently used by
FOSS projects such as Git, Tmux, Pry, Rspec) provides a default
location for various file formats, including config/rc files.

This comment refactors app_generator.rb to load railsrc from
XDG_CONFIG_HOME if both XDG_CONFIG_HOME is set and rails/railrc
exists within the XDG_CONFIG_HOME location.

To maintain backwards compatibility it defaults back to ~/.railsrc
if either XDG_CONFIG_HOME is not set or there is no rails/railsrc.

jonathanhefner commented on railties/lib/rails/generators/rails/app/app_generator.rb:
```diff
@@ -563,7 +563,15 @@ def prepare!
 
       def self.default_rc_file
-        File.expand_path("~/.railsrc")
+        return File.expand_path("~/.railsrc") unless ENV["XDG_CONFIG_HOME"]
+
+        xdg_railsrc = ENV["XDG_CONFIG_HOME"] + "/rails/railsrc"
+        rc_file = if File.exist?(xdg_railsrc)
+          xdg_railsrc
+        else
+          "~/.railsrc"
+        end
+        File.expand_path(rc_file)
```
I feel like this control flow is a little hard to follow.  What do you think about this arrangment:

```suggestion
        xdg_railsrc = File.expand_path("rails/railsrc", ENV["XDG_CONFIG_HOME"]) if ENV["XDG_CONFIG_HOME"]

        if xdg_railsrc && File.exist?(xdg_railsrc)
          xdg_railsrc
        else
          File.expand_path("~/.railsrc")
        end
```

quintrino replied:
@jonathanhefner Thanks!

Yep, I like the the flow and readability of this better, I've updated the commits now.  

I've updated the USAGE file as well. 

Hmm, I did a search for ~/.railsrc so I missed the File.join("~", ".railsrc"), good catch!

One option might be to extract the xdg detection logic into a shared parent module perhaps? 

We could do something like Rails.rc_file but I'm guessing the Rails module itself is fairly slow to change?

Once we decide on the location for storing the logic I'll attempt to write a test for it. 

Thanks for getting back to me. This is actually my first Open Source PR. 🙂

jonathanhefner reviewed (commented):
Cool!  What about a test for this?  And we should probably add mention to [USAGE](https://github.com/rails/rails/blob/42914820d91ded5b1fb285448bbde8620e00f33c/railties/lib/rails/generators/rails/app/USAGE#L6) too.

Also, we should probably do something with [this default](https://github.com/rails/rails/blob/42914820d91ded5b1fb285448bbde8620e00f33c/railties/lib/rails/commands/plugin/plugin_command.rb#L16) in `PluginCommand`, though I'm not sure of the best course of action there.

quintrino committed:
Refactor default_rc_file based on jonathanhefner PR feedback

quintrino committed:
Update generators/rails/app/USAGE to explain railsrc XDG compliance

jonathanhefner commented:
@quintrino I haven't forgotten about this!

> One option might be to extract the xdg detection logic into a shared parent module perhaps?

I think this is the correct answer, but, after some thinking, I want to factor out more than just the `default_rc_file` method.  For now, though, let's go with what we have.  This PR will add XDG support to `rails new`, and I'll work on a follow-up PR that adds XDG support to `rails plugin new`.

> Once we decide on the location for storing the logic I'll attempt to write a test for it.

:+1: For now, let's test `ARGVScrubber.default_rc_file`.

Oh, also, could you [add a changelog entry](https://guides.rubyonrails.org/contributing_to_ruby_on_rails.html#updating-the-changelog)?

> This is actually my first Open Source PR.

That's awesome!!  I hope for many more! :partying_face: :champagne:

sinsoku commented on railties/lib/rails/generators/rails/app/app_generator.rb:
```diff
@@ -563,7 +563,12 @@ def prepare!
 
       def self.default_rc_file
-        File.expand_path("~/.railsrc")
+        xdg_railsrc = File.expand_path("rails/railsrc", ENV["XDG_CONFIG_HOME"]) if ENV["XDG_CONFIG_HOME"]
+        if xdg_railsrc && File.exist?(xdg_railsrc)
```
According to the XDG Base Directory Specification, if `$XDG_CONFIG_HOME` is either not set or empty, you should use `$HOME/.config`.

> $XDG_CONFIG_HOME defines the base directory relative to which user specific configuration files should be stored. If $XDG_CONFIG_HOME is either not set or empty, a default equal to $HOME/.config should be used.
>
> ref: https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html

It would probably work with the following code.

```ruby
xdg_config_home = ENV["XDG_CONFIG_HOME"].presence || '~/.config'
xdg_railsrc = File.expand_path("rails/railsrc", xdg_config_home)
if File.exist?(xdg_railsrc)
  xdg_railsrc
else
  File.expand_path("~/.railsrc")
end
```

jonathanhefner replied:
@sinsoku I think you are correct in terms of the spec.  The downside of falling back to `~/.config` is that it makes this feature appear to be Linux-specific (even though it isn't), which makes it less likely to be merged.

On the other hand, both [Ruby](https://github.com/ruby/ruby/blob/b7e1de117e20911ed86b229a4c2f912e469c6472/lib/rubygems/defaults.rb#L128) and [git](https://git-scm.com/docs/git-config#Documentation/git-config.txt-XDGCONFIGHOMEgitconfig) fall back to `~/.config` regardless of OS, so I think there's an argument to be made for doing the same.

jonathanhefner commented:
@quintrino  Are you still interested in working on this?

quintrino committed:
Refactor default_rc_file based on sinsoku PR feedback

quintrino committed:
Add test for ARGVScrubber.default_rc_file method

fileutils is required because tmpdir does not allow you to create a new
temp directory with a set title 'inside' another temporary directory, a
file is created for railsrc because tempfile creates a custom and unique
name preventing the specification of a file called 'railsrc'

Currently this test is failing because the ARGVScrubber class seems to
be loading directly from the gem (which still has the previous code)
rather than from the file itself. (loading the code in a pry session
and using show-method ARGVScrubber.default_rc_file reveals it is coming
from
gems/railties-6.0.3.3/lib/rails/generators/rails/app/app_generator.rb:536
instead of
railties/lib/rails/generators/rails/app/app_generator.rb

quintrino commented:
Apologies on the delay on this, I was made redundant from my job 2 days after I committed this so this fell off my radar. 

@sinsoku I agree with the inclusion of ~/.config as a fallback, I've added in your changes now. 

@jonathanhefner I've created a test for ARGVScrubber.default_rc_file but at the moment it's failing locally for me. 
I've been using 
`ruby -I test railties/test/generators/argv_scrubber_test.rb` to run the test file. 

I suspect it could be because of something in $LOAD_PATH or possibly because the command I'm using isn't loading all of rails repo, and defaulting to the installed gems rather than my modified code. Any thoughts on what could be causing it to load from the gem rather than the app_generator.rb file? 

Happy to hear thoughts on how to improve the testing (I've listed some of the issues I ran into with testing the existence of an exact file in the last commit to explain why I wasn't able to use tempfile (as far as I can see).

sinsoku commented:
@quintrino You need to run the test in the `railties` directory.
I was able to run the test on my local machine, so I think it will probably work in your machine as well.

```console
$ cd railties
$ ruby -I test test/generators/argv_scrubber_test.rb
Run options: --seed 19374

# Running:

..........

Finished in 0.004136s, 2417.7950 runs/s, 5077.3695 assertions/s.
10 runs, 21 assertions, 0 failures, 0 errors, 0 skips
```

The test code is successful, but there are RuboCop errors.

You have to check the result of executing "RuboCop / build (pull_request)" at the bottom of this pull request and correct the code style.

sinsoku commented on railties/test/generators/argv_scrubber_test.rb:
```diff
@@ -56,6 +57,20 @@ def self.default_rc_file
+        railsrc.close
+
+        scrubber = Class.new(ARGVScrubber)
+        assert_equal scrubber.default_rc_file, railrc_path
```
It is better to clean the file as it still exists after running the test.

```ruby
  assert_equal scrubber.default_rc_file, railrc_path
ensure
  File.delete(railrc_path) if File.exist?(railrc_path)
end
```

quintrino replied:
I've added `File.delete` now.

quintrino committed:
Switch to double quote strings

quintrino committed:
Add file deletion for argv_scrubber test

rafaelfranca commented on railties/test/generators/argv_scrubber_test.rb:
```diff
@@ -56,6 +57,21 @@ def self.default_rc_file
+
+        scrubber = Class.new(ARGVScrubber)
+        assert_equal scrubber.default_rc_file, railrc_path
+        File.delete(railrc_path) if File.exist?(railrc_path)
```
Let's move to an `ensure` block and also unset the `ENV["XDG_CONFIG_HOME"]` value.

quintrino replied:
Thanks @rafaelfranca, I actually ended up switching to Jonathan Hefners code below, but good point on unsetting the ENV, I hadn't considered that.

jonathanhefner commented on railties/test/generators/argv_scrubber_test.rb:
```diff
@@ -56,6 +57,21 @@ def self.default_rc_file
         assert_equal [], args
       end
 
+      def test_xdg_config_no_custom_rc
+        xdg = Dir.tmpdir
+        ENV["XDG_CONFIG_HOME"] = xdg
+        rails_xdg_path = File.join(xdg, "rails")
+        Dir.mkdir(rails_xdg_path, 0700) unless Dir.exist?(rails_xdg_path)
+        Dir.new(rails_xdg_path)
+        railrc_path = File.join(rails_xdg_path, "railsrc")
+        railsrc = File.new(railrc_path, "w")
+        railsrc.close
+
+        scrubber = Class.new(ARGVScrubber)
+        assert_equal scrubber.default_rc_file, railrc_path
+        File.delete(railrc_path) if File.exist?(railrc_path)
```
@quintrino I hope you don't mind me piling on! :pray:  I felt like this test code could be a bit more focused:

```suggestion
      def test_default_rc_file_with_xdg_config_home
        Dir.mktmpdir do |dir|
          rc_file = File.join(dir, "rails/railsrc")
          FileUtils.mkdir_p(File.dirname(rc_file))
          FileUtils.touch(rc_file)

          switch_env("XDG_CONFIG_HOME", dir) do
            assert_equal rc_file, ARGVScrubber.default_rc_file
          end
        end
```

That will take care of cleaning up the file and the environment variable.  Though to use `switch_env` you'll need to add `require "env_helpers"` to the top of the file, and `include EnvHelpers` to the top of the test class ([similar to this](https://github.com/rails/rails/blob/1fd17336544a04119316c0a9dfc2b1464b4b3f30/railties/test/generators/actions_test.rb#L5-L9)).

quintrino replied:
@jonathanhefner I've applied these changes now, let me know if you think it needs anything else.

quintrino committed:
Refactor test_xdg_config_no_custom_rc with jonathanhefner feedback

quintrino committed:
Add Changelog for railsrc xdg-specification compliance

jonathanhefner reviewed (commented):
A couple of nitpicks.  Would you mind also [squashing the commits](https://guides.rubyonrails.org/contributing_to_ruby_on_rails.html#squashing-commits)?

quintrino committed:
Apply backtick and whitespace refactoring suggestsions from jonathanhefner

quintrino commented on railties/CHANGELOG.md:
```diff
@@ -1,3 +1,13 @@
+
+    Rails new will load the railsrc file from XDG_CONFIG_HOME/rails/railsrc
+    It will then fallback to .config if XDG_CONFIG_HOME not specified
+    It will then fallback to HOME/.railsrc
```
Added the blank line and backticks now, if everything looks good to you then I'll squash and force-push up everything in a single commit for review.

jonathanhefner commented on railties/CHANGELOG.md:
```diff
@@ -1,8 +1,8 @@
-    Rails new will load the railsrc file from XDG_CONFIG_HOME/rails/railsrc
-    It will then fallback to .config if XDG_CONFIG_HOME not specified
-    It will then fallback to HOME/.railsrc
+    Rails new will load the `railsrc` file from `XDG_CONFIG_HOME/rails/railsrc`.
+    It will then fallback to `.config` if `XDG_CONFIG_HOME` not specified.
+    It will then fallback to `HOME/.railsrc`.
```
"Rails new" could use some backticks too!  Also, to clarify what `.config` means, what do you think about:

```suggestion
    `rails new` will now look for the default `railsrc` file at
    `$XDG_CONFIG_HOME/rails/railsrc` (or `~/.config/rails/railsrc` if
    `XDG_CONFIG_HOME` is not set).  If this file does not exist, `rails new`
    will fall back to `~/.railsrc`.
```

quintrino replied:
Sure, I've made these changes now.

quintrino committed:
Apply jonathanhefner suggested Changelog refactoring
